source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2397 @@
+
+  - ``events``: Required array of documents. List of events, which are expected
+    to be observed (in this order) on the corresponding client while executing
+    `operations`_. If the array is empty, the test runner MUST assert that no
+    events were observed on the client.
+
+    The structure of each document is defined in `expectedEvent`_.
+
+.. _test_outcome:
+
+- ``outcome``: Optional array of documents. Each document will specify expected
+  contents of a collection after all operations have been executed. The list of
+  documents therein SHOULD be sorted ascendingly by the ``_id`` field to allow
+  for deterministic comparisons.
+
+  If set, the array should contain at least one document. The structure of each
+  document is defined in `collectionData`_.
+
+
+operation
+~~~~~~~~~
+
+An operation to be executed as part of the test.
+
+The structure of this document is as follows:
+
+.. _operation_name:
+
+- ``name``: Required string. Name of the operation (e.g. method) to perform on
+  the object.
+
+.. _operation_object:
+
+- ``object``: Required string. Name of the object on which to perform the
+  operation. This should correspond to either an `entity`_ name (for
+  `Entity Test Operations`_) or "testRunner" (for `Special Test Operations`_).
+  If the object is an entity, The YAML file SHOULD use an `alias node`_ for its
+  ``id`` field (e.g. ``object: *collection0``).
+
+.. _operation_arguments:
+
+- ``arguments``: Optional document. Map of parameter names and values for the
+  operation. The structure of this document will vary based on the operation.
+  See `Entity Test Operations`_ and `Special Test Operations`_.
+
+  The ``session`` parameter is handled specially (see `commonOptions_session`_).
+
+.. _operation_expectedError:
+
+- ``expectedError``: Optional document. One or more assertions for an expected
+  error raised by the operation. The structure of this document is
+  defined in `expectedError`_.
+
+  This field is mutually exclusive with
+  `expectedResult <operation_expectedResult_>`_ and
+  `saveResultAsEntity <operation_saveResultAsEntity_>`_.
+
+.. _operation_expectedResult:
+
+- ``expectedResult``: Optional mixed type. A value corresponding to the expected
+  result of the operation. This field may be a scalar value, a single document,
+  or an array of documents in the case of a multi-document read.  Test runners
+  MUST follow the rules in `Evaluating Matches`_ when processing this assertion.
+
+  This field is mutually exclusive with
+  `expectedError <operation_expectedError_>`_.
+
+.. _operation_saveResultAsEntity:
+
+- ``saveResultAsEntity``: Optional string. If specified, the actual result
+  returned by the operation (if any) will be saved with this name in the
+  `Entity Map`_.  The test runner MUST raise an error if the name is already in
+  use. If the operation does not return a value (e.g. void method), the test
+  runner MAY choose to store an empty value (e.g. ``null``) or do nothing and
```

Comment:
I think I would prefer to force the test runner to store an empty value (e.g. `null`, `None`) so that the key is defined. That would at least make a test file deterministic without considering runtime operation, since one could very well see that a key was referenced in `createEntities` and/or `saveResultAsEntity`. In such a case, I think it's more helpful for a test runner to raise an error that the key's type would be unexpected rather than the key itself being undefined.

As for how this comes up in practice, consider that `findAndModify` methods might return a document or empty value if nothing matches. I doubt there's any need to use `saveResultAsEntity` in that case (at least today), but I think it's worth clarifying any edge cases in the test runner from the get-go.

Reply:
Always storing a value regardless of whether it's considered "empty" sounds good to me. 👍